tool/test.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
+
+# Fast fail the script on failures.
+set -e
+
+# Write errors to stderr
+function error() {
+  echo "FAIL: $@" 1>&2
+}
+
+# Checks the formatting of the Dart files to make sure it conforms to dartfmt
+# output.
+function check_formatting() {
+  # Read files one line at a time, in case there are spaces in the names.
+  local fmt_result=()
+  IFS=$'\n' fmt_result=($("$dartfmt" -n "$@" 2>&1))
+  local fmt_code=$?
+  if [[ "$fmt_code" != 0 ]]; then
+    error "dartfmt exited with the following exit code: ${fmt_code}"
+    echo "${fmt_result[@]}" 1>&2
+    exit 1
+  elif [[ "${#fmt_result[@]}" == 0 ]]; then
+    echo "No formatting errors!"
+  else
+    error "There are formatting errors in the following files:"$'\n'
+    for file in "${fmt_result[@]}"; do
+      echo "===== $file =====" 1>&2
+      cp "$file" "$file.expected"
+      "$dartfmt" -w "$file.expected" &> /dev/null
+      echo "----- expected format -----" 1>&2
+      cat "$file.expected" 1>&2
+      echo "----- current format -----" 1>&2
+      cat "$file" 1>&2
+      echo "----- diff -----" 1>&2
+      diff "$file.expected" "$file" 1>&2 || true
+      echo "===== /end $file ====="$'\n' 1>&2
+      rm "$file.expected"
+    done
+    error "Formatting errors detected by dartfmt in ${#fmt_result[@]} dart file(s)."
+    return 1
+  fi
+}
+
+# Deploys website to Firebase, trying up to total_tries times.
+function deploy() {
+  local total_tries="$1"
+  local remaining_tries=$(($total_tries - 1))
+  local project="$2"
+  while [[ "$remaining_tries" > 0 ]]; do
+    # FIREBASE_TOKEN is set in the .cirrus.yml file.
+    firebase deploy --token "$FIREBASE_TOKEN" --non-interactive --project "$project" && break
+    remaining_tries=$(($remaining_tries - 1))
+    error "Error: Unable to deploy project $project to Firebase. Retrying in five seconds... ($remaining_tries tries left)"
+    sleep 5
+  done
+
+  [[ "$remaining_tries" == 0 ]] && {
+    error "Deployment still failed after $total_tries tries: '$@'"
+    return 1
+  }
+  return 0
+}
+
+# Use the version of Dart SDK from the Flutter repository instead of whatever
+# version is in the PATH. FLUTTER_ROOT comes from the Cirrus Docker image.
+export PATH="$FLUTTER_ROOT/bin/cache/dart-sdk/bin:$FLUTTER_ROOT/bin:$PATH"
```

Comment:
Yes, it does mean that for runs via Cirrus. I added this here because I was thinking of what happens if you run this script locally on your workstation (which should be a valid thing to do, assuming you have all the dependencies installed).

I did add something to automatically set FLUTTER_ROOT if it's unset, to make running it locally even easier.

Reply:
👍